win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+            {
+                LogError("Failure in CreateFileA, full_file_name=%s", full_file_name);
+                result = false;
+            }
+            else
+            {
+                /*Codes_SRS_FILE_WIN32_43_002: [ file_create shall call SetFileCompletionNotificationModes to disable calling the completion port when an async operations finishes synchrounously.]*/
+                if (!SetFileCompletionNotificationModes(handle->h_file, FILE_SKIP_COMPLETION_PORT_ON_SUCCESS))
+                {
+                    LogError("Failure in SetFileCompletionNotificationModes, full_file_name=%s", full_file_name);
+                    result = false;
+                }
+                else
+                {
+                    /*Codes_SRS_FILE_WIN32_43_003: [ file_create shall initialize a threadpool environment by calling InitializeThreadpolEnvironment.]*/
+                    InitializeThreadpoolEnvironment(&handle->cbe);
+
+                    /*Codes_SRS_FILE_WIN32_43_004: [ file_create shall obtain a PTP_POOL struct by calling execution_engine_win32_get_threadpool on execution_engine.]*/
+                    handle->ptp_pool = execution_engine_win32_get_threadpool(execution_engine);
+
+                    /*Codes_SRS_FILE_WIN32_43_005: [ file_create shall register the threadpool environment by calling SetThreadpoolCallbackPool on the initialized threadpool environment and the obtained ptp_pool ]*/
+                    SetThreadpoolCallbackPool(&handle->cbe, handle->ptp_pool);
+
+                    /*Codes_SRS_FILE_WIN32_43_006: [ file_create shall create a cleanup group by calling CreateThreadpoolCleanupGroup.]*/
+                    handle->ptp_cleanup_group = CreateThreadpoolCleanupGroup();
+
+                    if (handle->ptp_cleanup_group == NULL)
+                    {
+                        LogError("Failure in CreateThreadpoolCleanupGroup, full_file_name=%s", full_file_name);
+                        result = false;
+                    }
+                    else
+                    {
+                        /*Codes_SRS_FILE_WIN32_43_007: [ file_create shall register the cleanup group with the threadpool environment by calling SetThreadpoolCallbackCleanupGroup.]*/
+                        SetThreadpoolCallbackCleanupGroup(&handle->cbe, handle->ptp_cleanup_group, on_close_threadpool_group_member);
+
+                        /*Codes_SRS_FILE_WIN32_43_033: [ file_create shall create a threadpool io with the allocated FILE_HANDLE and on_file_io_complete_win32 as a callback by calling CreateThreadpoolIo]*/
+                        handle->ptp_io = CreateThreadpoolIo(handle->h_file, on_file_io_complete_win32, NULL, &handle->cbe);
+
+                        if (handle->ptp_io == NULL)
+                        {
+                            LogError("Failure in CreateThreadpoolIo, full_file_name=%s", full_file_name);
+                            result = false;
+                        }

```

Comment:
>} [](start = 24, length = 1)

style: Should have an else which should have the code from below:

```
    /*Codes_SRS_FILE_43_038: [ file_create shall register user_report_fault_callback with argument user_report_fault_context as the callback function to be called when the callback specified by the user for a specific asynchronous operation cannot be called. ]*/
    handle->user_report_fault_callback = user_report_fault_callback;
    handle->user_report_fault_context = user_report_fault_context;
    /*Codes_SRS_FILE_WIN32_43_009: [ file_create shall succeed and return a non - NULL value.]*/
```

Reply:
Move stuff from last few lines to in here.

---
In reply to: [456707239](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456707239) [](ancestors = 456707239)